hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/persistence/ContainerSchemaManager.java

```diff
@@ -0,0 +1,105 @@
+import org.hadoop.ozone.recon.schema.tables.daos.ContainerHistoryDao;
+import org.hadoop.ozone.recon.schema.tables.daos.MissingContainersDao;
+import org.hadoop.ozone.recon.schema.tables.pojos.ContainerHistory;
+import org.hadoop.ozone.recon.schema.tables.pojos.MissingContainers;
+import org.jooq.DSLContext;
+import org.jooq.Record2;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Provide a high level API to access the Container Schema.
+ */
+@Singleton
+public class ContainerSchemaManager {
+  private ContainerHistoryDao containerHistoryDao;
+  private MissingContainersDao missingContainersDao;
+  private ContainerSchemaDefinition containerSchemaDefinition;
+
+  @Inject
+  public ContainerSchemaManager(ContainerHistoryDao containerHistoryDao,
+              ContainerSchemaDefinition containerSchemaDefinition,
+              MissingContainersDao missingContainersDao) {
+    this.containerHistoryDao = containerHistoryDao;
+    this.missingContainersDao = missingContainersDao;
+    this.containerSchemaDefinition = containerSchemaDefinition;
+  }
+
+  public void addMissingContainer(long containerID, long time) {
+    MissingContainers record = new MissingContainers(containerID, time);
+    missingContainersDao.insert(record);
+  }
+
+  public List<MissingContainers> getAllMissingContainers() {
+    return missingContainersDao.findAll();
+  }
+
+  public boolean isMissingContainer(long containerID) {
+    return missingContainersDao.existsById(containerID);
+  }
+
+  public void deleteMissingContainer(long containerID) {
+    missingContainersDao.deleteById(containerID);
+  }
+
+  public void upsertContainerHistory(long containerID, String datanode,
+                                     long time) {
+    DSLContext ctx = containerSchemaDefinition.getDSLContext();
+    Record2<Long, String> record =
+        ctx.newRecord(
+        CONTAINER_HISTORY.CONTAINER_ID,
+        CONTAINER_HISTORY.DATANODE_HOST).value1(containerID).value2(datanode);
+    ContainerHistory newRecord = new ContainerHistory();
+    newRecord.setContainerId(containerID);
+    newRecord.setDatanodeHost(datanode);
+    newRecord.setLastReportTimestamp(time);
+    if (containerHistoryDao.existsById(record)) {
+      newRecord.setFirstReportTimestamp(
+          containerHistoryDao.findById(record).getFirstReportTimestamp()
+      );
+      containerHistoryDao.update(newRecord);
+    } else {
+      newRecord.setFirstReportTimestamp(time);
+      containerHistoryDao.insert(newRecord);
+    }
+  }
+
+  public List<ContainerHistory> getAllContainerHistory(long containerID) {
+    return containerHistoryDao.fetchByContainerId(containerID);
+  }
+
+  public List<ContainerHistory> getLatestContainerHistory(long containerID,
+                                                          int limit) {
+    // Get container history sorted in descending order of timestamp
+    List<ContainerHistory> containerHistory =
```

Comment:
Can we try and use ORDER BY and LIMIT on the DB side (using SELECT API maybe) instead of client side sorting and filtering?